pyvista/examples/_example_loader.py

```diff
@@ -0,0 +1,538 @@
+    def _filesize_format(self) -> _FilePropStrType_co:
+        """Return the formatted size of all file(s)."""
+
+    @property
+    @abstractmethod
+    def total_size(self) -> str:
+        """Return the total size of all files."""
+
+    @property
+    @abstractmethod
+    def reader(
+        self,
+    ) -> Optional[Union[pyvista.BaseReader, Tuple[Optional[pyvista.BaseReader], ...]]]:
+        """Return the base file reader(s) used to read the files."""
+
+
+@runtime_checkable
+class _Downloadable(Protocol[_FilePropStrType_co]):
+    """Class which implements a 'download' method."""
+
+    @abstractmethod
+    def download(self) -> _FilePropStrType_co:
+        """Download and return the file path(s)."""
+
+
+@runtime_checkable
+class _Loadable(Protocol):
+    """Class which implements a 'load' method."""
+
+    @abstractmethod
+    def load(self) -> Any:
+        """Load the dataset."""
+
+
+class _SingleFile(_FileProps[str, int]):
+    """Wrap a single file."""
+
+    def __init__(self, filename):
+        from pyvista.examples.downloads import USER_DATA_PATH
+
+        self._filename = (
+            filename if os.path.isabs(filename) else os.path.join(USER_DATA_PATH, filename)
+        )
+
+    @property
+    def filename(self) -> str:
+        return self._filename
+
+    @property
+    def _filesize_bytes(self) -> int:
+        return _get_file_or_folder_size(self.filename)
+
+    @property
+    def _filesize_format(self) -> str:
+        return _format_file_size(self._filesize_bytes)
+
+    @property
+    def total_size(self) -> str:
+        return self._filesize_format
+
+    @property
+    def reader(self) -> Optional[pyvista.BaseReader]:
+        return None
+
+
+class _SingleFileLoadable(_SingleFile, _Loadable):
+    """Wrap a single file for loading.
+
+    Specify the read function and/or load functions for reading and processing the
+    dataset. The read function is called on the filename first, then, if a load
+    function is specified, the load function is called on the output from the read
+    function.
+
+    Parameters
+    ----------
+    filename
+        Path of the file to be loaded.
+
+    read_func
+        Specify the function used to read the file. Defaults to :func:`pyvista.read`.
+        This can be used for customizing the reader's properties, or using another
+        read function (e.g. :func:`pyvista.read_texture` for textures). The function
+        must have the filename as the first argument and should return a dataset.
+        If default arguments are required by your desired read function, consider
+        using :class:`functools.partial` to pre-set the arguments before passing it
+        as an argument to the loader.
+
+    load_func
+        Specify the function used to load the file. Defaults to `None`. This is typically
+        used to specify any processing of the dataset after reading. The load function
+        typically will accept a dataset as an input and return a dataset.
```

Comment:
`read_func` should be changed to `reader`, and only allow setting to a `pyvista.BaseReader` class. Any custom file reader functions should be lumped into the `load_func` param instead. This way, instead of simply calling `pyvista.read` on the file, we can instead do `pyvista.get_reader` first in `__init__` and also set the reader as the loader's `reader` property. The default `read_func` would then become `self.reader.read()`.

Reply:
This is a non-trivial fix, and I'm not convinced the effort is entirely worth, especially if the current `_example_loader.py` api is kept fully private long-term.

The main motivation for having the `reader` property is actually to extract the reader's `type` as part of the metadata extraction. So, it doesn't really matter much if the `reader` instance isn't the actual reader used when `read_func` is called. So, perhaps we can simply make the property private (i.e. rename to  `_reader` and only implement the API change from `reader_func` -> `reader`  if there is a need for this feature in the future.

I can make the change of `reader` -> `_reader` in #5814.